react-js/react-components-and-hierarchies.md

```diff
@@ -0,0 +1,438 @@
+
+The difference between Components and built-in JSX tags is that components are created by the user while JSX tags are pre-defined
+
+Consider the following example
+```
+class App extends React.Component {
+   render() {
+     return <h1>Hello World  </h1>;
+  }
+}
+
+ReactDOM.render(
+  <App />,
+  document.getElementById('root')
+);
+```
+In this example, App is a component while h1 is a built-in JSX tags
+
+
+### How to pass data Between React Components
+
+1. Read [Passing Data Between React Components](https://medium.com/@ruthmpardee/passing-data-between-react-components-103ad82ebd17) 
+2. Read [React: Communication Between Components](https://blog.bitsrc.ioreact-communication-between-components-c0cfccfa996a)
+3. Watch [React components communication tutorial ](https://www.youtube.com/watch?v=dyL99ACQfsM)
+4. Watch [Pass data child to parent](https://www.youtube.com/watch?v=0FWrZF1qWfE)
+
+In react, sometimes we want components with communicate with each other. This communication can either be :
+1. From Parent to Child 
+2. From Child to Parent 
+
+#### From Parent to Child  — Use a prop
+One of the simplest and easiest ways to pass data to components is through props. Props are similar to a parameter in a function. If a parent component wants to feed data to a child component, it simply passes it via props
+*Example*
+
+```
+class Name extends React.Component {
+  render() {
+    return <h2>I am {this.props.brand}!</h2>;
+  }
+}
+
+class Person extends React.Component {
+  render() {
+    return (
+      <div>
+         <h1>What is your name</h1>
+         <Name brand="Annu" />
+      </div>
+    );
+  }
+}
+
+ReactDOM.render(<Person />, document.getElementById('root'));
+
+```
+[Run code](https://codepen.io/annu12340/pen/WNQGBjy?editors=1010)\
+Here we have created 2 components- Car and Garage. We send the "brand" property from the Garage component to the Car component using props
+
+
+
+#### From Child to Parent — Use a callback and states
+You may also want to pass data from child to parent. For this we can use states and callback methods
+Consider the 2 components - parent and child.
+Here we wish the child to pass a message 'Data received' to its parent, when a button is pressed
+*Example*
+
+Parent.js
+```
+
+class Parent extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      fromChild: "empty"
+    };
+  }
+
+  update = data => {
+    this.setState({ fromChild: data });
+  };
+
+  render() {
+    return (
+      <div className="App">
+        <h1>Parent Component </h1>
+        <p>
+          data coming from child<b>: {this.state.fromChild} </b>
+        </p>
+        <Child update={this.update} />
+      </div>
+    );
+  }
+}
+
+
+```
+Child.js
+```
+class Child extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      child: "Data received"
+    };
+  }
+
+  update = data => {
+    this.setState({ fromChild: data });
+  };
```

Comment:
Only the Parent should have state, to make this example easier to understand (and be more realistic)

Reply:
Please address this comment. The current example is more complicated than it needs to be and so obscures the lesson.

Reply:
‼️ This remains incomplete and is a very important change.